test/e2e/compact_test.go

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package e2e_test
+
+import (
+	"context"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/cortexproject/cortex/integration/e2e"
+	e2edb "github.com/cortexproject/cortex/integration/e2e/db"
+	"github.com/go-kit/kit/log"
+	"github.com/oklog/ulid"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	"github.com/thanos-io/thanos/pkg/block"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/objstore/client"
+	"github.com/thanos-io/thanos/pkg/objstore/s3"
+	"github.com/thanos-io/thanos/pkg/promclient"
+	"github.com/thanos-io/thanos/pkg/testutil"
+	"github.com/thanos-io/thanos/pkg/testutil/e2eutil"
+	"github.com/thanos-io/thanos/test/e2e/e2ethanos"
+)
+
+func TestCompact(t *testing.T) {
+	t.Parallel()
+	l := log.NewLogfmtLogger(os.Stdout)
+
+	// blockDesc describes a recipe to generate blocks from the given series and external labels.
+	type blockDesc struct {
+		series                   []labels.Labels
+		extLsets                 []labels.Labels
+		numberOfSamplesPerSeries int
+	}
+
+	for i, tcase := range []struct {
+		name          string
+		blocks        []blockDesc
+		replicaLabels []string
+		query         string
+
+		expected               []model.Metric
+		numberOfModifiedBlocks float64
+		numberOfBlocks         uint64
+		numberOfSamples        uint64
+		numberOfSeries         uint64
+		numberOfChunks         uint64
+	}{
+		{
+			name: "overlapping blocks with matching replica labels",
+			blocks: []blockDesc{
+				{
+					series: []labels.Labels{labels.FromStrings("a", "1", "b", "2")},
+					extLsets: []labels.Labels{
```

Comment:
I would make it more tricky. Those labels sorted, add more labels than just one. Add one block that is totally irrevant to this replica. Add one block with just `"ext1", "value1" ?